network/local.go

```diff
@@ -0,0 +1,487 @@
+package network
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+)
+
+// NewLocalRouter returns a fresh router which uses only local queues. It uses
+// the default local manager.
+// If you need multiple independent local-queues, use NewLocalRouterWithManager.
+// In case of an error it is returned together with a nil-Router.
+func NewLocalRouter(sid *ServerIdentity) (*Router, error) {
+	return NewLocalRouterWithManager(defaultLocalManager, sid)
+}
+
+// NewLocalRouterWithManager is the same as NewLocalRouter but takes a specific
+// LocalManager. This is useful to run parallel different local overlays.
+// In case of an error it is returned together with a nil-Router.
+func NewLocalRouterWithManager(lm *LocalManager, sid *ServerIdentity) (*Router, error) {
+	h, err := NewLocalHostWithManager(lm, sid.Address)
+	if err != nil {
+		return nil, err
+	}
+	return NewRouter(sid, h), nil
+}
+
+// LocalManager keeps a reference to all opened local connections.
+// It also keeps track of who is "listening", so it's possible to mimic
+// Conn & Listener.
+type LocalManager struct {
+	// queues maps a remote endpoint to its packet queue. It's the main
+	// structure used to communicate.
+	queues map[endpoint]*connQueue
+	sync.Mutex
+	// The listening-functions used when a new connection-request arrives.
+	listening map[Address]func(Conn)
+
+	// connection-counter for giving unique IDs to each connection.
+	counter uint64
+}
+
+// NewLocalManager returns a fresh new manager that can be used by LocalConn,
+// LocalListener & LocalHost.
+func NewLocalManager() *LocalManager {
+	return &LocalManager{
+		queues:    make(map[endpoint]*connQueue),
+		listening: make(map[Address]func(Conn)),
+	}
+}
+
+// defaultLocalManager can be used if you need only one LocalManager.
+var defaultLocalManager = NewLocalManager()
+
+// endpoint represents one endpoint of a connection.
+type endpoint struct {
+	// addr is the Address of this endpoint.
+	addr Address
+	// uid is a unique identifier of the remote endpoint
+	// it's unique  for each direction:
+	// 127.0.0.1:2000 -> 127.0.0.1:2000 => 14
+	// 127.0.0.1:2000 <- 127.0.0.1:2000 => 15
```

Comment:
Yes it's to handle one very specific case regarding Local conns: you can listen on `127.0.0.1:2000`, and when you want to connect to it, you must pass one `from Address` and one `to Address`. Nothing rejects you from selecting the same address.